sql_test.go

```diff
@@ -318,3 +319,47 @@ func testNullUUIDUnmarshalJSONMalformed(t *testing.T) {
 		t.Fatal("json.Unmarshal err = <nil>, want error")
 	}
 }
+
+func BenchmarkNullMarshalJSON(b *testing.B) {
+	b.Run("Valid", func(b *testing.B) {
+		u, err := FromString("6ba7b810-9dad-11d1-80b4-00c04fd430c8")
+		if err != nil {
+			b.Fatal(err)
+		}
+		n := NullUUID{UUID: u, Valid: true}
+		for i := 0; i < b.N; i++ {
+			n.MarshalJSON()
```

Comment:
Benchmarking gotcha: the compiler is allowed to optimize this call away since nothing is using the return values.

Reply:
I don't believe that is true. This benchmark takes 72ns so it's clearly calling `MarshalJSON` and the Go standard library rarely use the result [`strings/strings_test.go#L332-L350)`](https://github.com/golang/go/blob/9f0234214473dfb785a5ad84a8fc62a6a395cbc3/src/strings/strings_test.go#L332-L350).

Reply:
Absolutely and this would actually be pretty performant if implemented with a generic function that `Parse` and `UnmarshalText` would delegate to (at the cost of dropping support for Go 1.17 and less).

Reply:
Maybe, but it would be a breaking change since it would invalidate a large number of existing benchmarks. So if introduced it would likely exempt benchmarks (and tests) from any such optimization (you have to do the same when writing a benchmark suite in C/C++ where the compilers are much more aggressive `#pragma GCC ...`).

Reply:
I'm basing this on the _A note on compiler optimizations_ section of [this blog](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go) from Dave Cheney.  The compiler choosing to eliminate that call is within the spec and, since the toolchain isn't covered by the Go1 compatibility guarantee, it wouldn't even be considered breaking.

Reply:
Yep, you're entirely correct the spec does not cover the toolchain and TIL in some cases the compiler will elide the function call unless the result it used (I checked this using a simple `square(x int) int` function, which being "pure" in the C sense is probably why the compiler decided it could be elided - though interestingly it does not elide the loop).

I'm happy to change this to assign to a dummy variable.